[ShapeUp] improve UX when adding/editing storage

lorenzo-cavazzi opened the pull request:
This PR fixes a few issues when adding/editing storage:
- Prevents a bunch of default values to be set automatically, especially when many advanced options exist.
- Correct use of placeholders for default values and examples.
- Remove values correctly when editing an existing storage.
- Handles non-stringy values correctly when listing the storage options in the storage list.
- Use `<datalist>` to pick values from the examples list.
    - Mind that examples from the backend are not 100% reliable; we can't use a `<select>` here, and I would rather keep a simple solution than mess up with autocomplete examples and "use-react-form." Support for the datalist tag is pretty good on Firefox, while the UX on Chrome is still not that good, especially for longer lists.

/deploy renku=rclone-sessions renku-notebooks=rclone-csi amalthea=main renku-data-services=main extra-values=notebooks.cloudstorage.enabled=true,notebooks.cloudstorage.storageClass=csi-rclone-test,notebooks.cloudstorage.installRCloneCSI=false

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -608,6 +565,81 @@ function SecretOptionWarning({
+  const [isTruncated, setIsTruncated] = useState(true);
+  const contentRef = useRef<HTMLDivElement>(null);
+
+  const toggleTruncate = () => {
+    setIsTruncated(!isTruncated);
+  };
```
Prefer `useCallback()` version:

```suggestion
  const toggleTruncate = useCallback(() => {
    setIsTruncated((isTruncated) => !isTruncated);
  }, []);
```

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -608,6 +565,81 @@ function SecretOptionWarning({
+  return (
+    <div>
+      <div
+        className={cx("overflow-hidden", "local-style")}
```
What's `"local-style"`?

lorenzo-cavazzi replied:
Ouch, that's a leftover from a previous test

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -648,6 +680,7 @@ function AddStorageOptions({
+  // ? We treat checkbox and password separately, while the rest is an input
   const optionItems =
     options &&
     options.map((o) => {
```
This is difficult to read, consider moving this to components (e.g. `<CheckboxOptionItem>`, `<PasswordOptionItem>`, `<DropdownOptionItem>`, etc.).

lorenzo-cavazzi replied:
That's a good point; I split that long chunk of code into multiple components

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -683,7 +704,7 @@ function AddStorageOptions({
+            {o.friendlyName ?? o.name} {warning}
           </label>
 
           {inputType === "checkbox" ? (
```
Checkboxes should be placed to the left of their label, not to the right.

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -751,35 +772,64 @@ function AddStorageOptions({
           )}
-          <div className={cx("form-text", "text-muted")}>{o.help}</div>
+          <div className={cx("form-text", "text-muted")}>
+            {/* {o.help} */}
```
```suggestion
```

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -816,6 +866,10 @@ function AddStorageOptions({
 
   const sourcePath = (
     <div className="mb-3">
+      <Label className="form-label" for="add-storage-name">
```
Wrong `for=` here.

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -608,6 +565,81 @@ function SecretOptionWarning({
+  examples: CloudStorageSchemaOptionExample[];
+  listId: string;
+}
+function OptionDataList({ examples, listId }: OptionDataListProps) {
```
Note: the same functionality can be achieved with `react-autosuggest`.

lorenzo-cavazzi commented:
@leafty thank you for the review. I addressed your comments in ea48f28c400cfc9dee8614c01df04e39ca941646

There are a couple of additional commits that add minimal changes:
- Hide a couple of storage services that we don't want to support in the UI, at least initially.
- Restore the correct tag when invoking renku actions

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -343,6 +347,246 @@ function AddStorageAdvanced({ storage, setStorage }: AddStorageStepProps) {
+  onFieldValueChange,
+  option,
+}: InputOptionItemProps) {
+  const additionalProps: Record<string, string> = {};
+  if (inputType === "dropdown") {
+    additionalProps.list = `${option.name}__list`;
+  }
```
```suggestion
  const additionalProps: Record<string, string> = {
    ...(inputType === "dropdown" ? { list: `${option.name}__list` } : {}),
  };
```

leafty reviewed (approved):
LGTM :100: just some small things

lorenzo-cavazzi commented:
@leafty fixes from the last review are in this commit https://github.com/SwissDataScienceCenter/renku-ui/pull/2962/commits/0a995c9717ac8a0de739a85d72fdac53d7e7bfbb

leafty commented on client/src/features/project/components/cloudStorage/AddOrEditCloudStorage.tsx:
```diff
@@ -343,6 +347,246 @@ function AddStorageAdvanced({ storage, setStorage }: AddStorageStepProps) {
+}
+
+interface CheckboxOptionItemProps {
+  control: Control<FieldValues, any>; // eslint-disable-line @typescript-eslint/no-explicit-any
```
I think we can force this to be `void` if we call `useForm<Values, void>({<defaultValues>})` when using `react-hook-form`. The `context: TContext` field is used only if we use a third-party library for field validation.

`unknown` can work too I think. We don't need to address the use of `any` in this PR anyway.
